Introduce `UK_LLSYSCALL_R_U_DEFINE` and register `clone` with it

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`, `arm64`
 - Platform(s): `kvm`
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Some runtimes assume that the child of `clone` inherits full context of the parent (except for scratch registers).
Introduce `struct uk_syscall_regs`:
```C
struct uk_syscall_regs {                                                                                                                                                      
       struct __regs regs;                                                                                                                                                    
       struct ukarch_ulctx ulctx;                                                                                                                                             
       __u8 ectx[UKARCH_ECTX_SAVE_MAX_SIZE];                                                                                                                                  
       __u8 pad[UK_SYSCALL_PAD_SIZE];                                                                                                                                         
};
```
A structure made up of architecture specific structures:
- `struct __regs` for general purpose registers
- new `struct ukarch_ulctx`, meant to help the kernel keep track of certain special registers an application may change and which we may want to know of in case we have to swap them with ours when entering a system call. E.g. TLS pointer in bincompat, or the `x86_64` `gs_base` register which an application may change through `arch_prctl` system call.
- `ectx` mean to represent the architecture specific slot where we may want to save/restore to/from when enterin/exiting a system call

Introduce `UK_LLSYSCALL_R_U_DEFINE`, a system call registration macro alternative that offers a system call access to the caller/parent's full `struct uk_syscall_regs` context. Two system calls shall be registered: `clone` and [`arch_prctl`](https://github.com/unikraft/app-elfloader/pull/65). Additionally, `uk_syscall6_r` and `ukplat_syscall_handler` shall now make use of this generic structure.

Thus, now `clone` is able to make the child inherit full context of the parent if desired and satisfy picky runtimes.

Depends on #1173 and #1174

mogasergiu committed:
lib/syscall_shim: Introduce `struct ukarch_sysregs`

Define an architecture specific userland context that an architecture
can use to store things it may want to preserve or switch between
when entering/exiting Unikraft context during normal userland
execution (e.g. running in conjunction with binary syscalls).

For now, this is defined as a struct with a base field for both
ARM64 and x86_64, namely the `tpidr_el0` and `fs_base` registers
respectively.

Additionally, x86_64 has one more specific field:
- `gs_base`, that refers to the value of the `gs_base` register of
the application before calling the system call

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Add `ukarch_sysregs` from/to Unikraft

Implement switch from/to functions to be used in conjunction
with the userland context.

The switch to Unikraft operation assumes that it can only be called
from within Unikraft context (i.e. not directly by the app) and, as
such, the app's `gs_base` register value was preserved within
`X86_MSR_KERNEL_GS_BASE` following a `swapgs` on syscall entry.
This value will be thus stored in the current userland
context.

The switch to userland operation assumes the same thing and, therefore,
it will set `X86_MSR_GS_BASE` to the current `struct lcpu` and
`X86_MSR_KERNEL_GS_BASE` to the app's preserved `gs_base` register
within the userland context of the app's thread we are currently
switching to.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/arm64: Add `sysregs` switch from/to Unikraft

Implement `switch_ul`/`switch_uk` functions to be used in conjunction
with the system registers. Useful when wanting to ensure consistency
between register states of the application versus Unikraft core's.

This will also switch the current TLS pointer to the one that was
stored in the userland context.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/posix-process: Copy full parent registers context on `clone`

Register `clone` through `UK_LLSYSCALL_R_U_DEFINE` to have full
access to the `struct uk_syscall_ctx` context of the caller/parent.
This way we will be able to give the child the same exact register
context as the parent, except, of course, for the obvious registers
(e.g. `%rax` equal to `0` on `x86_64` and `x0` equal to `0`
on `ARM64`).

Some runtimes, such as the Go one, may expect some registers
to be exact the same as those of the parent.

Registers that may not be preserved may be the scratch ones
(e.g. `%r11` on x86).

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Add `fs_base` get/set in `sysregs`

Add basic setter/getter operations for the `fs_base` field of the
userland context structure that is meant to represent the userspace
TLS.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/arm64: Add `tpidr_el0` get/set for `sysregs`

Add basic setter/getter operations for the `tpidr_el0` field of the
userland context structure that is meant to represent the userspace
TLS.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Implement TLS switch from/to Unikraft

Implement two basic methods:
- TLS switch to Unikraft: stores current userland TLS into `fs_base`
fielf of the userland context and updates the active TLS to that of
Unikraft
- TLS switcho to userland: undoes what switchon did, by setting the
active TLS as the one that was previously stored in `fs_base`

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/arm64: Implement TLS switch from/to Unikraft

Implement two basic methods:
- TLS switch to Unikraft: stores current userland TLS into `tpidr_el0`
fielf of the userland context and updates the active TLS to that of
Unikraft
- TLS switch from Unikraft: undoes what the previous switch did, by
setting the active TLS as the one that was previously stored in
`tpidr_el0`

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Introduce `struct uk_syscall_ctx`

Define a new structure, `struct uk_syscall_ctx`, which is currently
composed of the following fields:
- a `struct __regs regs` field meant to contain the register context
of the userland caller before executing the `syscall` instruction
- a `struct ukarch_sysregs sysregs` field meant to contain the architecture
specific context containing only the registers that have to be preserved,
kept track of and switched between during system call entry/exit.
- a slot big enough to hold the saving of the architecture specific
extended context
- padding to ensure that the structure is aligned end-to-end, w.r.t. the
alignment of the extended context saving area.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
include/essentials: Provide `__naked` macro

Add a macro for `__attribute__((naked))`. This hints to the compiler
to generate code without prolog and epilog code. This comes in handy
when writing inline assembly functions only. Although caution may
be required as the ABI must still be respected and the caller
registers that are to be preserved by the callee must still be
preserved.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Add `gs_base` getter/setter

Implement basic methods for getting/setting the `gs_base` field of
`struct ukarch_sysregs`.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
arch/x86/x86_64: Allow asm files access to register offsets macros

It may come in handy to be able to know the offsets of each register
inside the `struct __regs` structure during more fragile states of
execution, usually coded in assembly.

Therefore, move the guarding `#ifdef` below the register definitions
so that assembly files may be able to include this header and benefit
from the aforementioned macros.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
arch/x86/x86_64: Add constant assertions to register structures

Allow catching missized formats of the `struct __regs` at build time
by using `UK_CTASSERT` on them.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Obsolete `uk_syscall_r_fn`

This is no longer used and unnecessarily adds a bit of difficulty
when it comes to adding new system call registration macros. Thus,
remove it.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86: Add `uk_syscall_ctx` saving prologue

Implement an x86 inline assembly macro `UK_SYSCALL_USC_PROLOGUE_DEFINE`
function named after `pname` that switches to the auxiliary stack,
and starts pushing and storing the caller's context before passing it
as argument to a function whose name is defined by the second argument,
`fname`.

This function shall be defined as a `__naked` function, meaning the
compiler does not provide us withABI compliant prologue/epilogue.
Althrough we supposedly don't touch the callee preserved registers in
this function, fully optimized images may inline the `fname` function
and end up messing the ABI. Therefore, make double sure this is avoided
by restoring the callee preserved registers before switching back to the
initial stack and returning.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/arm64: Add `uk_syscall_ctx` saving prologue

Implement an ARM inline assembly macro `UK_SYSCALL_USC_PROLOGUE_DEFINE`
function named after `pname` that switches to the auxiliary stack,
and starts pushing and storing the caller's context before passing it
as argument to a function whose name is defined by the second argument,
`fname`.

Althrough we supposedly don't touch the callee preserved registers in
this function, fully optimized images may inline the `fname` function
and end up messing the ABI. Therefore, make double sure this is avoided
by restoring the callee preserved registers before switching back to the
initial stack and returning.

To further emphasize to the compiler that we do not want anything
else in this function other than the prologue, optimize it with O3.

Notice that we make use of `TPIDRRO_EL0` as a scratch register so that
we can temporarily preserve x0's value. This should be safe as the
application/user themselves cannot change this register's value,
which means this register's value will always be known by us and we
could restore it anytime.

Some example OSes making use of TPIDRRO_EL0 would include Zephyr using
use this register for holding its `struct __cpu`, Windows using it to
hold the current CPU number, or Linux which uses it as both a scratch
register and a secondary thread ID.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/x86_64: Add `struct uk_syscall_ctx` restore

Introduce an assembly written method `uk_syscall_ctx_popall` that
may be used in conjunction with a context that may be saved through a
`struct uk_syscall_ctx` structure. This method assumes that `%rsp`
points to the aforementioned structure and, furthermore, makes the
assumption that it is already aligned in memory to what this structure
should be (i.e. alignment of ECTX). Furthermore, and most importantly,
this method assumes that `gs_base` is set to the current LCPU's
correspondent `struct lcpu` pointer as it ends in a `swapgs`.

`uk_syscall_ctx_popall` simply disables IRQ's and restores the
extended context, userland context (`gs_base`/`fs_base` at the moment)
and the general purpose registers only to then finish with a `swapgs`
to swap Unikraft/Application `gs_base`s and an `iretq` to "teleport"
in the expected context.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim/arch/arm64: Add `struct uk_syscall_ctx` restore

Introduce an assembly written method `uk_syscall_ctx_popall` that
may be used in conjunction with a context that may be saved through a
`struct uk_syscall_ctx` structure. This method assumes that `sp`
points to the aforementioned structure and, furthermore, makes the
assumption that it is already aligned in memory to what this structure
should be (i.e. alignment of ECTX).

`uk_syscall_ctx_popall` simply disables IRQ's and restores the
extended context, userland context (`tpidr_el0` only at the moment)
and the general purpose registers only to `eret` to "teleport"
to the expected context.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Define `UK_LLSYSCALL_R_U_DEFINE`

Implement `UK_LLSYSCALL_R_U_DEFINE`, a `struct uk_syscall_ctx`
alternative to `UK_LLSYSCALL_R_DEFINE`. System call that get
registered through this macro shall have an inline assembly
preamble call them. This preamble will preserve the
`struct uk_syscall_ctx` of the caller and pass it to the
registered syscall.

The system call function will be declared with the registered
arguments and, additionally, with a hidden `struct uk_syscall_ctx *usr`
argument that will contain the aforementioned saved context. This
is very useful when it comes to system calls such as `fork` or `clone`
that may want to duplicate the caller's/parent's context into the child.

E.g.
Consider the usual registration with `UK_LLSYSCALL_R_DEFINE`:
```C
UK_LLSYSCALL_R_DEFINE(type, sysname, type0, arg0, type1, arg1)
{
	...
}
```
This will end up creating two global scope symbols:
```
uk_syscall_e_sysname
uk_syscall_r_sysname
```
These symbols will end up calling a statically declared
`__uk_syscall_r_sysname`/`__uk_syscall_e_sysname` behind the scenes
with the necessary arguments and this will be the actual function
executing the code between the brackets `{ }`.

Now consider the new variant, `UK_LLSYSCALL_R_U_DEFINE`:
```C
UK_LLSYSCALL_R_U_DEFINE(type, sysname, type0, arg0, type1, arg1)
{
	...
}
```
Behind the scenes this will create four global scope symbols:
```
uk_syscall_e_sysname
uk_syscall_r_sysname
uk_syscall_e_u_sysname
uk_syscall_r_u_sysname
```
`uk_syscall_r_sysname` and `uk_syscall_e_sysname` will actually be
the assembly preamble that stores the caller's context in the form of
`sturct uk_syscall_ctx` and then they call `uk_syscall_r_u_sysname`
and `uk_syscall_e_u_sysname` respectively, which represent the actual
code written between brackets `{ }`. These, just like before, will call
`__uk_syscall_r_u_sysname`/`__uk_syscall_e_u_sysname`, which represent
a key useful difference from their alternative without the `_u`:
- The `__uk_syscall_r_sysname`/`__uk_syscall_e_sysname` will have the
following signature:
```C
static inline
type __uk_syscall_r_sysname(type0 arg0, type1 arg1);
static inline
type __uk_syscall_r_sysname(type0 arg0, type1 arg1);
```
- The `__uk_syscall_r_u_sysname`/`__uk_syscall_e_u_sysname` will have the
following signature:
```C
static inline __used
type __uk_syscall_r_u_sysname(struct uk_syscall_ctx *usr,
			      type0 arg0, type1 arg1);
static inline __used
type __uk_syscall_r_u_sysname(struct uk_syscall_ctx *usr,
			      type0 arg0, type1 arg1);
```
Thus, the programmer can use the `_U_` system call registration variant
if they require access to the full context of the caller/parent.

Finally, add equivalent `ARG_MAP`/`PRINTD` and the like macro's to
accomodate this new type of registration.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Save `struct uk_syscall_ctx` on binary syscalls

Make it so that `ukplat_syscall_handler` now receives a
`struct uk_syscall_ctx` pointer as its argument instead.

This structure shall be saved on the stack by the early assembly
entry for each architecture. Furthermore, prefer `sysregs` API
over other methods of touching the context.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Save `struct uk_syscall_ctx` in `uk_syscall6_r`

Being the center of all system call dispatching, it may be useful
to let `uk_syscall6_r` receive a `struct uk_syscall_ctx` as its
argument, containing the full context of the caller/parent.

Thus, if a system call has its `_u_` variant defined, that will be
the one that is used instead and the argument passed to it is
the aforementioned `struct uk_syscall_ctx` structure.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/syscall_shim: Obsolete `_uk_syscall_[return_addr|ultlsp]`

Now that we rely fully on `struct uk_syscall_ctx` and
`struct ukarch_sysregs`, we no longer need these TLS variables.

Therefore, remove them and any references of them from the codebase.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented on lib/syscall_shim/arch/arm64/syscall_regs.S:
```diff
@@ -0,0 +1,59 @@
+	/* Restore x18, x19 */
+	ldp	x18, x19, [x19, #16 * 9]
+
+	msr	daifclr, #2
```
Remembered about this while working on smth else, but basically this is not needed because `eret` will restore corresponding `PSTATE`.

TODO: Delete this line

michpappas commented on lib/syscall_shim/arch/arm64/include/uk/ulctx.h:
```diff
@@ -0,0 +1,20 @@
+struct ukarch_ulctx {
+	__uptr tpidr_el0;
+
+	__u8 pad[8];	/* Make sure we are a multiple of 16 bytes */
```
Perhaps instead of padding you can do:
```c
struct ukarch_ulctx {
 ...
} __align(__STACK_ALIGN_SIZE);
```

michpappas commented on include/uk/essentials.h:
```diff
@@ -105,6 +105,9 @@ extern "C" {
 #define __fallthrough          __attribute__((fallthrough))
 #endif
+#ifndef __naked
+#define __naked                __attribute__((naked))
```
If `__attribute__((naked))` is not supported by `arm64`, then perhaps this should be conditionally defined to `x86_64` or to `!arm64`, or at least add a comment for people who are not aware?

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -399,6 +399,107 @@ typedef long uk_syscall_arg_t;
+#endif /* !LCPU_AUXSP_OFFSET */
+
+#if defined(__X86_64__)
+#define UK_SYSCALL_USR_PROLOGUE_DEFINE(pname, fname, x, ...)		\
```
This should be defined in an architecture-specific header. Also, since this looks a bit strange, could you please provide some context on why this defined in a header as a macro and not in a separate assembly file?

mogasergiu replied:
This needs to be expanded to a very specific function signature whose inline assembly calls a very specific function w.r.t. to the arguments passed to the macro. Each system call registered through this macro will have its own personalized defined prologue.

michpappas commented on include/uk/essentials.h:
```diff
@@ -357,6 +357,10 @@ extern "C" {
 #define UK_NARGS(...) \
 	__UK_NARGS_X(, ##__VA_ARGS__, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, \
 		     15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
+#define UK_NACTUAL_ARGS(...) \
```
This is a bit too tailored to `syscall_shim` to belong to `essentials.h` IMO. A comment block that provides some context would also be helpful.

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -213,6 +213,73 @@ typedef long uk_syscall_arg_t;
+/* Taken from regmap_linuxabi.h, but defined differently so as not to
+ * needlessly contaminate the namespace of source files including this header
+ */
+#if (defined __X86_64__)
```
Similarly to other cases, please use architecture-specific headers.

michpappas commented on lib/syscall_shim/arch/x86_64/ulctx.c:
```diff
@@ -0,0 +1,32 @@
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+#include <uk/arch/ctx.h>
```
These look like a lot of headers for these two functions, I would expect you'd only need around half of them.

michpappas commented on lib/syscall_shim/arch/x86_64/include/uk/ulctx.h:
```diff
@@ -25,6 +25,10 @@ UK_CTASSERT(sizeof(struct ukarch_ulctx) == UKARCH_ULCTX_SIZE);
 
 void ukarch_ulctx_set_tlsp(struct ukarch_ulctx *ulctx, __uptr tlsp);
+
+void ukarch_ulctx_switchoff_tls(struct ukarch_ulctx *ulctx);
```
I'm not 100% keen on these names, as they imply that the TLS is something you toggle on and off 💡, while what we actually do is switch between user / kernel TLS. Besides, it's easy to forget after a while which one is `on` and which one is `off`. Something along the lines of `ukarch_ulctx_switch_ultls()` / `ukarch_ulctx_switch_uktls()` would be more clear IMO.

Edit: After seeing `ukarch_ulctx_switchon()` / `ukarch_ulctx_switchoff()` I get the idea. Still, perhaps it would be a bit more clear to use `_ultls` instead of `_tls`? If everyone thinks they're clear, I don't have a strong opinion though.

michpappas commented on lib/syscall_shim/arch/x86_64/ulctx.c:
```diff
@@ -29,4 +29,27 @@ void ukarch_ulctx_set_tlsp(struct ukarch_ulctx *ulctx, __uptr tlsp)
+{
+	struct uk_thread *t = uk_thread_current();
+
+	UK_ASSERT(ulctx);
```
It would be useful if we somehow could assert that we are currently in `switchon` state. I don't have a good idea on how though 🤔

mogasergiu replied:
I used to have these assertions but they are not quite compatible with the native use-case where the GS register pair is equal :(. But I could add the `IS_LCPU_PTR` :+1:

michpappas commented on lib/syscall_shim/arch/x86_64/ulctx.c:
```diff
@@ -10,8 +10,42 @@
+	UK_ASSERT(ulctx);
+	UK_ASSERT(lcpu_get_current());
+
+	/* This can only be called from Unikraft ctx in bincompat mode.
```
Similarly to `ukarch_ulctx_switchon_tls()`, an assertion that we're operating in Unikraft context would be useful. Similarly to `ukarch_ulctx_switchon_tls()`, I have no good ideas to propose 🤔 

I would also add that it's not enough to be in Unikraft context, but also in syscall / switchon context.

michpappas commented on lib/syscall_shim/arch/x86_64/ulctx.c:
```diff
@@ -10,8 +10,42 @@
+void ukarch_ulctx_switchon(struct ukarch_ulctx *ulctx)
+{
+	UK_ASSERT(ulctx);
+	UK_ASSERT(lcpu_get_current());
```
Maybe leave a comment on what does this protect from? (also use the `IS_LCPU_PTR`)

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -35,18 +35,66 @@
 
+#include <uk/ulctx.h>
+
+#if defined(__X86_64__)
```
These should be in architecture-specific headers (eg `lib/syscall-shim/arch/x86_64/include/uk/arch/syscall.h`). That way you also don't have to deal with the "Undefined architecture" problem.

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -35,18 +35,66 @@
+
+#if defined(__X86_64__)
+/* 512 for the actual structure + 64 bytes for the XSAVE header */
+#define UKARCH_ECTX_SAVE_MAX_SIZE			578
```
512 + 64 = 576, not 578 🤔

mogasergiu replied:
:nerd_face: :computer: :brain: :mage_man:

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -35,18 +35,66 @@
 extern "C" {
 #endif
 
+struct uk_syscall_regs {
```
The naming here is a bit messed up 😵‍💫 
- This struct contains the entire cpu context, ie more than regs, yet it's named `regs`. Since context normally refers to the entire CPU state, so that would be a more appropriate name for this structure.
- For what we call context in this PS (which is not really context - only `FS`/`GS`), we use `ulctx`, yet for the extended context we call it just `ectx`.

~~But if you guys are happy with it, I don't have a strong opinion at all :)~~ Edit: Actually the more I see it, the more I don't like it, because it makes calling code unintuitive.

mogasergiu replied:
I shall introduce a nerw structure `struct ukarch_actx`, which will include `(__regs, s/ulctx/{fs/gs + tpidr_el0/pad}, ectx)` and all `*ctx*` stuff shall go intro `include/uk/arch/actx.h`

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -399,6 +399,107 @@ typedef long uk_syscall_arg_t;
+
+#if defined(__X86_64__)
+#define UK_SYSCALL_USR_PROLOGUE_DEFINE(pname, fname, x, ...)		\
+	long __used __naked __attribute__((optimize("O3")))		\
```
Shouldn't that be `O0` instead?

mogasergiu replied:
Will keep `O3` on ARM and naked only on x86 :+1:. Also add a comment. All I need is for the ABI compliant prologue/epilogue to not be generated and, unfortuantely, ARM does not support `naked`

michpappas reviewed (changes requested):
@mogasergiu amazing as always 😎 I am leaving some first comments. I might provide additional feedback after porting my work. Thanks!

PS: As with #1174 perhaps we can move the Arm stuff to a separate PR 🙏🏼

michpappas commented on lib/posix-process/clone.c:
```diff
@@ -305,6 +307,101 @@ static void _clone_child_gc(struct uk_thread *t)
 
+void uk_syscall_regs_popall(void);
+
+#if defined(__X86_64__)
```
Please move these into `lib/posix-process/arch/...`

michpappas commented on lib/posix-process/clone.c:
```diff
@@ -376,20 +472,20 @@ static int _clone(struct clone_args *cl_args, size_t cl_args_len,
+		uk_pr_debug(" child_tid: %p\n", (void *)cl_args->child_tid);
+	uk_pr_debug(" stack: %p\n", (void *)cl_args->stack);
+	uk_pr_debug(" tls: %p\n", (void *)cl_args->tls);
+	uk_pr_debug(" <return>: %p\n", (void *)usr->rip);
```
Should be: usr->regs.rip

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -351,6 +367,250 @@ typedef long uk_syscall_arg_t;
+		}							\
+		return ret;						\
+	}								\
+	static inline rtype __##rname(struct uk_syscall_regs *usr,	\
```
This one is problematic in syscalls with no args like `vfork`, as it evaluates to:
```
__##rname(struct uk_syscall_regs *usr, )
```
To reproduce:
```
UK_LLSYSCALL_R_U_DEFINE(pid_t, vfork)
```

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -61,110 +109,79 @@ extern "C" {
+#error "Missing register mappings for selected target architecture"
+#endif  /* !__X86_64__ && !__ARM_64__ */
+
+#define UK_USR_MAP2_1(m, type, arg) (long)usr->regs.usr_arg0
```
`UK_LLSYSCALL_R_U_DEFINE` with 1 param don't work because of `UK_USR_MAP2_7` missing.

To reproduce:
```
UK_LLSYSCALL_R_U_DEFINE(int, mysyscall, unsigned long, arg)
```

michpappas commented on lib/syscall_shim/include/uk/syscall.h:
```diff
@@ -399,6 +399,107 @@ typedef long uk_syscall_arg_t;
+		" * [ 2: 2]: Table Indicator - GDT, table 0\n\t"	\
+		" * [ 1: 0]: Requestor Privilege Level - ring 0\n\t"	\
+		" */\n\t"						\
+		"pushq	$(0x10)\n\t"					\
```
This could be a `#define` instead of magic. Same for line 448.

mogasergiu replied:
Sadly, just like other magic values for ARM, the headers defining the macro's for these magic values would introduce even more dependencies for `syscall_shim`, which, in turn, more dependencies for others that include `syscall.h` (which are many). Therefore, I chose to add very descriptive comments instead :disappointed:

mogasergiu commented:
@michpappas @skuenzer Ok, all comments should've been addressed.
One thing left to do: renamings?
- `struct uk_syscall_regs` -> `struct uk_syscall_ctx`? Keep in mind this will mean also changing all abbreviations from `USR`/`usr` to `USC`/`usc`.
- maybe `struct ukarch_ulctx`? idk

mogasergiu commented:
Another force-push to fix an implcit declaration warning of `clone_setup_child_ctx`

michpappas commented on include/uk/essentials.h:
```diff
@@ -360,6 +360,10 @@ extern "C" {
 #define UK_NARGS(...) \
 	__UK_NARGS_X(, ##__VA_ARGS__, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, \
 		     15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0)
+#define UK_NARGS_HALF(...) \
```
It would be useful to write a comment to explain why is that macro required and how it's used.

Also IIRC in the last review round it was agreed that this should be in `libsyscall_shim` as it's specific to that library. Does the change in name change that? Are there any other use-cases? If the answer is no then perhaps both `UK_NARGS` and `UK_NARGS_HALF` belong to `libyscall_shim`? @skuenzer any thoughts?

skuenzer replied:
`UK_NARGS` is generic enough to keep it here. It is just a way for using the pre-processor to count the number of arguments that a variadic macro got handed over. This information can be used to build more complex macros on top. `UK_NARGS_HALF` looks to me syscall_shim specific because I remember we had a macro like this originally in syscall_shim but I was able to rewrite the macros in syscall_shim to base them also on `UK_NARGS`. Is it really needed for your case?

mogasergiu replied:
I use it for `UK_USR_CALLMAPx`. Thought I would pass the actual number of arguments.. but I guess I could just use the double of that and just pretend I am using the actual number of args but I will have to have a look again because I don't remember much.

michpappas commented on plat/common/arm/traps_arm64.c:
```diff
@@ -224,13 +225,13 @@ void trap_el1_irq(struct __regs *regs)
 	struct ukarch_trap_ctx *ctx = (struct ukarch_trap_ctx *)data;
 
-	ukplat_syscall_handler(ctx->regs);
+	ukplat_syscall_handler((struct uk_syscall_regs *)ctx->regs);
```
Although not introduced in this PR, I'm not entirely happy with `ukarch_trap_ctx`. We have too many definitions of context (also `ukarch_ctx`). But let it be for now, we can look at it later 🙂

skuenzer replied:
Hum... I agree with this point that we have too many `ctx`es, but we probably need to clean this up later. What we also should avoid is to introduce contexts that are basically the same thing and we do need to do typecasting of the same pointer.
I wonder if `struct ukarch_trap_ctx` could be handed over directly to `ukplat_syscall_handler` and if this makes sense also for x86. :thinking: What do we have more on `ukarch_trap_ctx`?

mogasergiu replied:
**What do we have less**:
```C
struct ukarch_trap_ctx {
	struct __regs *regs;
	int trapnr;
	int error_code;

	unsigned long fault_address; /* for page faults */
};
```
```C
struct uk_syscall_regs {
	struct __regs regs;
	struct ukarch_ulctx ulctx;
	__u8 ectx[UKARCH_ECTX_SAVE_MAX_SIZE];
	__u8 pad[UK_SYSCALL_PAD_SIZE];
};
```

michpappas reviewed (changes requested):
@mogasergiu thanks for the updates! Besides the few minor comments below I think there are only a couple of naming issues left to go.

For `switchon` / `switchoff` I would propose something less ambiguous. One idea that comes to mind is `ukarch_ulctx_switch_ul` / `ukarch_ulctx_switch_uk` (userland / unikraft) or similar.

Also less important:
- `struct uk_syscall_regs`: This is actually a context, as by context one normally refers to the collective CPU state (yes _technically speaking_ it does consist of registers so _technically speaking_ the current name is also correct). One idea is to name it `struct uk_syscall_context` if it stays in the syscall shim, or perhaps more correctly make it generic, call it `uk_lcpu_context` and move it into `lcpu`. Or leave it as is.
- `struct __regs` -> `struct uk_lcpu_regs`. A bit of a misnomer as the rest of the struct is also registers, but we can keep it. Alternatively `struct uk_lcpu_gpregs` (general purpose regs). Or leave it as is and we rename it along with everything else once we migrate into `libuklcpu`.
- `struct ukarch_ulctx`: Should probably be `struct uk_lcpu_sysregs` or something. I think this one should actually be renamed in this PR, as `ctx` is definitely misused here.

mogasergiu commented:
Changelog

e493bd78177e39fc05654bb283a4c46c0ee8f933
arch/arm/arm64: Add constant assertions to register structures
- squashed into 765f4155890bafd6d3db2d673634b9c744200275 of #1174

9c940ab4bb91a1f904fc1045ffb9c7633945b5ef
lib/syscall_shim/arch/x86_64: Add fs_base getter/setter for ulctx
fdd52f147d98434b6ceb9fdadaaada7c8b1f013b
lib/syscall_shim/arch/arm64: Add tpidr_el0 getter/setter for ulctx
cc4a48aa37260bf03936856ff7ddda1a4c95a7c2
lib/syscall_shim/arch/x86_64: Add gs_base getter/setter
- s/ulctx/sysregs
- s/ULCTX/SYSREGS

44b13c828dbfa18606b15c94405d273b145e7b25
lib/syscall_shim/arch/x86_64: Implement TLS switchon/switchoff
e90361bfe78d00c097ecd3655b55b79aa2124f41
lib/syscall_shim/arch/arm64: Implement TLS switchon/switchoff
390cf661e698ce729be93c96c4af19bb9334eab4
lib/syscall_shim/arch/x86_64: Add ulctx switchon/switchoff
4fcbd9c73eca35ad7b97cc7789c310be02291664
lib/syscall_shim/arch/arm64: Add ulctx switchon/switchoff
- s/ulctx/sysregs
- s/ULCTX/SYSREGS
- switchon/switchoff -> switch_ul/switch_uk

006a5db79d1303a8cc2b55fd3be0c568eb94c914
lib/syscall_shim: Introduce struct uk_syscall_regs
- s/UK_SYSCALL_REGS*/UK_SYSCALL_CTX
- s/uk_syscall_regs/uk_syscall_ctx
- `UKARCH_ECTX_SAVE_MAX_SIZE` -> `UKARCH_ECTX_SIZE`
- `UKARCH_ECTX_ALIGN` -> `UKARCH_ECTX_ALIGN`
- s/ulctx/sysregs
- s/ULCTX/SYSREGS

3da0f6bc95c531e5ace6324eef1f50111f6a7ce3
lib/syscall_shim/arch/x86: Add uk_syscall_regs saving prologue
a3490253d5e1f6bf367b6bfd8c4bc89b2020c4cd
lib/syscall_shim/arch/arm64: Add uk_syscall_regs saving prologue
- s/UK_SYSCALL_REGS*/UK_SYSCALL_CTX
- s/uk_syscall_regs/uk_syscall_ctx
- s/ulctx/sysregs
- s/ULCTX/SYSREGS
- s/USR/USC
- switchon/switchoff -> switch_ul/switch_uk

962b429f6c4765691149d53d86e90c994d04fd7c
lib/syscall_shim/arch/x86_64: Add struct uk_syscall_regs restore
e60f0a96828e6b46ab3fbbe099a91948d4c5014f
lib/syscall_shim/arch/arm64: Add struct uk_syscall_regs restore
- s/UK_SYSCALL_REGS*/UK_SYSCALL_CTX
- s/uk_syscall_regs/uk_syscall_ctx
- s/ulctx/sysregs
- s/ULCTX/SYSREGS
- switchon/switchoff -> switch_ul/switch_uk

1352cfc6621837601f80fd81a1528dab02c3b330
lib/syscall_shim: Define UK_LLSYSCALL_R_U_DEFINE
20881b33784986d9d40fa2260b81f1b991d8ce75
lib/syscall_shim: Save struct uk_syscall_regs on binary syscalls
47da6d8ef5fd831de6d1efb6dea254ec25f7fd24
lib/syscall_shim: Save struct uk_syscall_regs in uk_syscall6_r
be9245a7a6673bca4cf81159e48eeccc195fa933
lib/posix-process: Copy full parent registers context on clone
- s/UK_SYSCALL_REGS*/UK_SYSCALL_CTX
- s/uk_syscall_regs/uk_syscall_ctx
- s/ulctx/sysregs
- s/ULCTX/SYSREGS
- switchon/switchoff -> switch_ul/switch_uk
- s/USR/USC

Deleted
bd70beecbf8362d8c6353dc8d33e351694cf669b
include/essentials: Define UK_NARGS_HALF macro
and doubled suffixes of UK_USC_EMAP/UK_USC_CALLMAP instead

mogasergiu commented:
Also moved https://github.com/unikraft/unikraft/commit/371d7030739d268c36bae14d09f8fd9f7461edea from https://github.com/unikraft/unikraft/pull/1175 to #1174 so that #1174 builds individually.

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

@mogasergiu once again thanks a lot for going through the length process for this merge!
@razvand this needs to be rebased and tested together with dependencies before it can be merged.

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

Socratesa commented on plat/kvm/arm/exceptions.S:
```diff
@@ -112,18 +113,27 @@
+	str	x0, [sp, #__SP_OFFSET]  /* Store old SP in auxiliary stack */
+	b	1f
 0:
 	/* Restore x0 */
```
Why not restore x0 in the case of a system call

mogasergiu replied:
Yeah you're right, x0 should have been restored on syscalls as well of course. I must have somehow wrongly changed this during the review process and not notice. The restoration of x0 should have been placed after the `1:` label instead.

razvand replied:
@mogasergiu , would this require an update (PR) from your side?

mogasergiu replied:
Yes I could look into it after I return from my vacation, unless someone else would like to have a go at it as well. Though the fix should be rather simple: moving the restoration of x0 two lines below should suffice if I see this right.

mogasergiu replied:
Done #1256
